php/PaypalIPN.php

```diff
@@ -0,0 +1,135 @@
+            return self::SANDBOX_VERIFY_URI;
+        } else {
+            return self::VERIFY_URI;
+        }
+    }
+
+
+    /**
+     * Verification Function
+     * Sends the incoming post data back to paypal using the cURL library.
+     *
+     * @return bool
+     * @throws Exception
+     */
+    function verifyIPN()
+    {
+        if ( ! count($_POST)) {
+            throw new Exception("Missing POST Data");
+        }
+
+        $raw_post_data = file_get_contents('php://input');
+        $raw_post_array = explode('&', $raw_post_data);
+        $myPost = [];
+        foreach ($raw_post_array as $keyval) {
+            $keyval = explode('=', $keyval);
+            if (count($keyval) == 2) {
+                // Since we do not want the plus in the datetime string to be encoded to a space, we manually encode it.
+                if ($keyval[0] === 'payment_date') {
+                    if (substr_count($keyval[1], '+') === 1) {
+                        $keyval[1] = str_replace('+', '%2B', $keyval[1]);
+                    }
+                }
+                $myPost[$keyval[0]] = urldecode($keyval[1]);
+            }
+        }
+
+        // Build the body of the verification post request, adding the _notify-validate command.
+        $req = 'cmd=_notify-validate';
+        $get_magic_quotes_exists = false;
+        if (function_exists('get_magic_quotes_gpc')) {
+            $get_magic_quotes_exists = true;
+        }
+        foreach ($myPost as $key => $value) {
+            if ($get_magic_quotes_exists == true && get_magic_quotes_gpc() == 1) {
+                $value = urlencode(stripslashes($value));
+            } else {
+                $value = urlencode($value);
+            }
+            $req .= "&$key=$value";
+        }
+
+        // Post the data back to paypal, using curl. Throw exceptions if errors occur.
+        $ch = curl_init($this->getPaypalUri());
+        curl_setopt($ch, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
+        curl_setopt($ch, CURLOPT_POST, 1);
+        curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+        curl_setopt($ch, CURLOPT_POSTFIELDS, $req);
+        curl_setopt($ch, CURLOPT_SSLVERSION, 6);
+        curl_setopt($ch, CURLOPT_SSL_VERIFYPEER, 1);
+        curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 2);
+
+        // This is often required if the server is missing a global cert bundle, or is using an outdated one.
+        if ($this->use_local_certs) {
+            curl_setopt($ch, CURLOPT_CAINFO, __DIR__ . "/cert/cacert.pem");
+        }
+        curl_setopt($ch, CURLOPT_FORBID_REUSE, 1);
+        curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 30);
+        curl_setopt($ch, CURLOPT_HTTPHEADER, ['Connection: Close']);
+        $res = curl_exec($ch);
+        $info = curl_getinfo($ch);
+        $http_code = $info['http_code'];
+
+        if ($http_code != 200) {
+            throw new Exception("PayPal responded with http code $http_code");
+        }
+
+        if ( ! ($res)) {
```

Comment:
We can pick the version you have as long as it is consistent across the project. 👍 . No changes necessary in this case.